Copilot CLI `--disable-mcp-server` / `--disable-builtin-mcps` not projectable from CopilotSdkOptions

**Reported by**: RevoBot .NET daemon team (GAP-9)
**Affected**: `AchieveAi.LmDotnetTools.CopilotSdkProvider` 1.0.32 and earlier
**Severity**: Medium — blocks consumers from suppressing a broken user-level MCP server in Copilot CLI sessions

## Summary

`CopilotAcpTransport` hardcodes the spawned Copilot CLI's argument list to `["--acp", "--stdio"]` at [`src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs:105`](https://github.com/achieveai/LmDotnetTools/blob/main/src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs#L105). There is no caller-controllable extension point for adding flags like `--disable-mcp-server <name>` or `--disable-builtin-mcps`.

Both flags are documented in `copilot --help` and verified working in `--acp --stdio` mode by direct repro. Consumers need them to suppress broken-by-design user-level MCP entries that Copilot CLI loads on every session (e.g. a user-level `github` HTTP MCP whose Bearer header references `\${VAR}` which Copilot does NOT expand, producing a noisy 400 on every session).

## Requested change

Add a new option on `CopilotSdkOptions`:

\`\`\`csharp
/// <summary>
/// MCP server names the spawned Copilot CLI should treat as disabled —
/// projected onto the CLI as one \`--disable-mcp-server <name>\` flag per
/// entry. Combines orthogonally with the per-session <see cref=\"McpServers\"/>
/// dictionary: servers in this list are excluded from Copilot's MCP
/// inventory regardless of whether they were configured at the user,
/// workspace, plugin, or session level.
/// </summary>
public IReadOnlyList<string> DisabledMcpServers { get; init; } = Array.Empty<string>();
\`\`\`

And project each entry onto the spawned CLI's argv in `CopilotAcpTransport.StartAsync`:

\`\`\`csharp
var argList = new List<string> { \"--acp\", \"--stdio\" };
foreach (var server in _options.DisabledMcpServers ?? Array.Empty<string>())
{
    if (!string.IsNullOrWhiteSpace(server))
    {
        argList.Add(\"--disable-mcp-server\");
        argList.Add(server);
    }
}
\`\`\`

Optional companion (less critical but symmetric):

\`\`\`csharp
/// <summary>
/// When true, the spawned Copilot CLI is launched with
/// \`--disable-builtin-mcps\`, suppressing Copilot's built-in
/// \`github-mcp-server\` (the read-only remote at
/// \`api.individual.githubcopilot.com/mcp/readonly\`). Use when the caller
/// is providing a fully-featured replacement.
/// </summary>
public bool DisableBuiltinMcps { get; init; } = false;
\`\`\`

Projected as:

\`\`\`csharp
if (_options.DisableBuiltinMcps)
{
    argList.Add(\"--disable-builtin-mcps\");
}
\`\`\`

## Empirical evidence

Probed Copilot CLI directly (no LmDotnetTools, no daemon — just a Node script driving a full ACP handshake against `copilot --acp --stdio`) on 2026-05-21 with CLI version `1.0.52.0`.

**Baseline (no disable flag)** — broken user-level `github` HTTP MCP fails with:
\`\`\`
[ERROR] Failed to start MCP client for remote server github:
    Streamable HTTP error: Error POSTing to endpoint: bad request:
    Authorization header is badly formatted
\`\`\`

**With `--disable-builtin-mcps`** — built-in `github-mcp-server` correctly suppressed; user-level `github` still loads (and still fails — the flag only affects built-ins).

**With `--disable-mcp-server github`** — user-level `github` HTTP entry silently skipped; clean log:
\`\`\`
[ERROR] Skipping disabled MCP server: github
\`\`\`

Both flags accepted by the CLI in `--acp --stdio` mode; both verified working at the protocol layer (full ACP `initialize` + `session/new` handshake completes, no JSON-RPC errors).

## Acceptance criteria

1. `CopilotSdkOptions` exposes `DisabledMcpServers : IReadOnlyList<string>` with a default of `Array.Empty<string>()`.
2. `CopilotAcpTransport.StartAsync` appends one `--disable-mcp-server <name>` pair to the CLI argv for each non-empty entry.
3. (Optional) `CopilotSdkOptions` exposes `DisableBuiltinMcps : bool` with a default of `false`; when true, `--disable-builtin-mcps` is appended.
4. Existing callers without either option set continue to launch with the original `[\"--acp\", \"--stdio\"]` argv (backward compatibility).
5. Test: `CopilotAcpTransportLaunchContractTests` (or equivalent) asserts the projected argv contains the expected flag entries.

## Downstream test waiting on this

RevoBot ships a `[Fact(Skip = \"Pending upstream …\")]` test `CopilotSdkOptions_DisabledMcpServers_field_exists_for_GAP_9` that uses reflection to query for the field. When the upstream field lands and the LmDotnetTools NuGet bumps, the Skip is removed and the test will verify the field's presence — a follow-up RevoBot fix will then plumb `[\"github\"]` through to suppress the broken user-level entry.

--- a/src/CopilotSdkProvider/Configuration/CopilotSdkOptions.cs
+++ b/src/CopilotSdkProvider/Configuration/CopilotSdkOptions.cs
@@ -113,6 +113,20 @@ public record CopilotSdkOptions
         = ImmutableDictionary<string, McpServerConfig>.Empty;
 
     /// <summary>
+    ///     Names of MCP servers the Copilot CLI should be told to disable. Each
+    ///     non-empty entry projects to one <c>--disable-mcp-server &lt;name&gt;</c>
+    ///     argument on the spawned CLI process. Null or whitespace entries are
+    ///     skipped silently. Order is preserved.
+    /// </summary>
+    public IReadOnlyList<string> DisabledMcpServers { get; init; } = [];
+
+    /// <summary>
+    ///     When true, the Copilot CLI is launched with <c>--disable-builtin-mcps</c>,
+    ///     suppressing the CLI's bundled built-in MCP servers. Default <c>false</c>.
+    /// </summary>
+    public bool DisableBuiltinMcps { get; init; }
+
+    /// <summary>
     ///     Pluggable launcher used to spawn the Copilot CLI process. Defaults to
     ///     <see cref="DefaultProcessLauncher.Instance"/> which executes the CLI
     ///     directly on the host. Inject a custom <see cref="IProcessLauncher"/>

--- a/src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs
+++ b/src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs
@@ -102,7 +102,7 @@ internal sealed class CopilotAcpTransport : IAsyncDisposable
             {
                 Agent = CliAgentKind.Copilot,
                 ExecutableHint = _options.CopilotCliPath,
-                Arguments = ["--acp", "--stdio"],
+                Arguments = BuildCliArguments(_options),
                 WorkingDirectory = workingDirectory,
                 EnvironmentOverrides = envOverrides,
                 StandardOutputEncoding = Encoding.UTF8,
@@ -670,6 +670,33 @@ internal sealed class CopilotAcpTransport : IAsyncDisposable
                 pending.TrySetException(ex);
             }
         }
+    }
+
+    private static IReadOnlyList<string> BuildCliArguments(CopilotSdkOptions options)
+    {
+        var args = new List<string>(capacity: 4) { "--acp", "--stdio" };
+
+        var disabled = options.DisabledMcpServers;
+        if (disabled != null)
+        {
+            foreach (var name in disabled)
+            {
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    continue;
+                }
+
+                args.Add("--disable-mcp-server");
+                args.Add(name);
+            }
+        }
+
+        if (options.DisableBuiltinMcps)
+        {
+            args.Add("--disable-builtin-mcps");
+        }
+
+        return args;
     }
 
     private static void WriteArbitraryValue(Utf8JsonWriter writer, object value, JsonSerializerOptions jsonOptions)

--- a/src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs
+++ b/src/CopilotSdkProvider/Transport/CopilotAcpTransport.cs
@@ -687,7 +687,7 @@ internal sealed class CopilotAcpTransport : IAsyncDisposable
                 }
 
                 args.Add("--disable-mcp-server");
-                args.Add(name);
+                args.Add(name.Trim());
             }
         }
 

--- a/tests/CopilotSdkProvider.Tests/Transport/CopilotAcpTransportLaunchContractTests.cs
+++ b/tests/CopilotSdkProvider.Tests/Transport/CopilotAcpTransportLaunchContractTests.cs
@@ -132,6 +132,24 @@ public class CopilotAcpTransportLaunchContractTests
     }
 
     [Fact]
+    public async Task StartAsync_DisabledServerName_IsTrimmed()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisabledMcpServers = ["  playwright  ", "\tmemory\n"],
+        });
+
+        request.Arguments.Should().Equal(
+            "--acp",
+            "--stdio",
+            "--disable-mcp-server",
+            "playwright",
+            "--disable-mcp-server",
+            "memory");
+    }
+
+    [Fact]
     public async Task StartAsync_DisableBuiltinMcps_EmitsBuiltinFlag()
     {
         var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions

--- a/tests/CopilotSdkProvider.Tests/Transport/CopilotAcpTransportLaunchContractTests.cs
+++ b/tests/CopilotSdkProvider.Tests/Transport/CopilotAcpTransportLaunchContractTests.cs
@@ -65,4 +65,131 @@ public class CopilotAcpTransportLaunchContractTests
         request.HostPaths.Should().ContainSingle()
             .Which.Should().BeEquivalentTo(new HostPathReference(workingDir, HostPathKind.WorkingDirectory));
     }
+
+    [Fact]
+    public async Task StartAsync_DefaultOptions_DoesNotEmitDisableMcpFlags()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+        });
+
+        request.Arguments.Should().Equal("--acp", "--stdio");
+    }
+
+    [Fact]
+    public async Task StartAsync_SingleDisabledMcpServer_EmitsOneDisableMcpServerFlag()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisabledMcpServers = ["playwright"],
+        });
+
+        request.Arguments.Should().Equal(
+            "--acp",
+            "--stdio",
+            "--disable-mcp-server",
+            "playwright");
+    }
+
+    [Fact]
+    public async Task StartAsync_MultipleDisabledMcpServers_EmitsRepeatedFlagsInOrder()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisabledMcpServers = ["playwright", "memory", "github"],
+        });
+
+        request.Arguments.Should().Equal(
+            "--acp",
+            "--stdio",
+            "--disable-mcp-server",
+            "playwright",
+            "--disable-mcp-server",
+            "memory",
+            "--disable-mcp-server",
+            "github");
+    }
+
+    [Fact]
+    public async Task StartAsync_NullOrWhitespaceServerName_IsSkipped()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisabledMcpServers = ["playwright", "  ", null!, string.Empty, "memory"],
+        });
+
+        request.Arguments.Should().Equal(
+            "--acp",
+            "--stdio",
+            "--disable-mcp-server",
+            "playwright",
+            "--disable-mcp-server",
+            "memory");
+    }
+
+    [Fact]
+    public async Task StartAsync_DisableBuiltinMcps_EmitsBuiltinFlag()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisableBuiltinMcps = true,
+        });
+
+        request.Arguments.Should().Equal("--acp", "--stdio", "--disable-builtin-mcps");
+    }
+
+    [Fact]
+    public async Task StartAsync_BothDisableFlags_EmitsDisabledServersBeforeBuiltinFlag()
+    {
+        var request = await CaptureLaunchRequestAsync(new CopilotSdkOptions
+        {
+            CopilotCliPath = "copilot-cli-mock",
+            DisabledMcpServers = ["playwright", "memory"],
+            DisableBuiltinMcps = true,
+        });
+
+        request.Arguments.Should().Equal(
+            "--acp",
+            "--stdio",
+            "--disable-mcp-server",
+            "playwright",
+            "--disable-mcp-server",
+            "memory",
+            "--disable-builtin-mcps");
+    }
+
+    private static async Task<ProcessLaunchRequest> CaptureLaunchRequestAsync(CopilotSdkOptions baseOptions)
+    {
+        var recorder = new RecordingLauncher();
+        var options = baseOptions with { ProcessLauncher = recorder };
+
+        await using var transport = new CopilotAcpTransport(options);
+
+        var workingDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+        Directory.CreateDirectory(workingDir);
+        try
+        {
+            var act = async () => await transport.StartAsync(
+                workingDir,
+                apiKey: null,
+                baseUrl: null,
+                requestHandler: (_, _, _) => Task.FromResult(default(JsonElement)),
+                notificationHandler: (_, _) => { },
+                ct: CancellationToken.None);
+
+            await act.Should().ThrowAsync<InvalidOperationException>();
+        }
+        finally
+        {
+            Directory.Delete(workingDir, recursive: true);
+        }
+
+        recorder.LastRequest.Should().NotBeNull();
+        return recorder.LastRequest!;
+    }
 }

